### ISSUE TITLE
Having Issues with Training

### ISSUE THREAD

**Issue by varunsharma38** [ofirnachum/tree_rnn#1]
Hi

I've been trying to work with your code. But I ran into a problem when I tried running sentiment.py for obtaining results over SST. The loss returned by the following statement:

loss, pred_y = model.train_step(tree, None)                 --line 99

is coming out to be **NaN**. Any help on this issue would be appreciated.
Also if you could explain the usage of the project(which files to run, any pre-requisites etc.)

Thanks in anticipation.

Varun

**Comment by ofirnachum** [comment]
Have you gone through the instructions in the readme (downloading and preparing the data)?

> Code for evaluation on the Standford Sentiment Treebank (used by the paper) is also available in sentiment.py. To run this, you'll need to download the relevant data. You can do so by cloning the treelstm git repo (https://github.com/stanfordnlp/treelstm) and running it's initial fetch_and_preprocess.sh script. After this, you'll also need to prepare the Glove vectors as .npy files using the data_utils.read_embeddings_into_numpy method.

**Comment by jainvidit** [comment]
We have followed all the steps given by you.
The problem comes out as stated by @varunsharma38 is
The `train_step` function is returning the loss as `NaN` and most of the values returned in `pred_y` are also `NaN`

**Comment by varunsharma38** [comment]
I have done the following things @ofirnachum :

Clone https://github.com/stanfordnlp/treelstm
Run fetch_and_preprocess.sh
Call data_utils.read_embeddings_into_numpy(glove.6B.vocab)
Run sentiment.py from the terminal
PS: I'm working with the Glove 6B dataset
Also there were some other issues with array broadcast and the glove.npy file was having a blank array because it won't convert strings to float

Hope that this explains the problem..
It would be really helpful if you could enumerate the steps required to run the code.

Thanks

**Comment by ofirnachum** [comment]
Ok - a couple questions:
1. Do the other demos work without issue?  E.g. modulo_demo.py?
2. Does the sentiment demo work without tunable word embeddings?  (Try commenting out lines 71-80.)
3. Does the sentiment demo work if you set the learning rate to 0?
4. Does the sentiment demo work if you set the learning rate to something small >0 (e.g. 0.0001)?

**Comment by varunsharma38** [comment]
modulo_demo.py works fine but simple_demo.py returns the following error:

![screenshot from 2016-05-28 11 44 44](https://cloud.githubusercontent.com/assets/6450451/15626118/d94b831c-24d9-11e6-935b-d0570571c282.png)

The Sentiment demo still return NaN upon setting the learning rate to 0 and 0.0000001

However, the sentiment demo works fine without tunable word embeddings. I **## had modified the code** to the following

# Tunable Word Embeddings:

`glove_vecs = []
 glove_words, glove_word2idx = [],[]
 embeddings = model.embeddings.get_value()
 print 'Reading Glove'
 glove_vecs = np.load(os.path.join(GLOVE_DIR, 'glove.npy'))
 glove_words = np.load(os.path.join(GLOVE_DIR, 'words.npy'))
 glove_word2idx = dict((word, i) for i, word in enumerate(glove_words))
 for i, word in enumerate(vocab.words):
      if word in glove_word2idx:
            #print word,glove_word2idx[word],glove_vecs[glove_word2idx[word]]
            a = glove_vecs[glove_word2idx[word]].copy()
            embeddings[i] = a.resize(300)
    model.embeddings.set_value(embeddings)`

# data_utils.read_embeddings_into_numpy:

`words = []
 array = []
 with open(file_name, 'r') as in_file:
        for line in in_file:
            fields = line.strip().split('\n')
        word = fields[0]
            if vocab and word not in vocab.word2idx:
                continue
            embedding = np.array( [[float(ord(f)) for f in field] for field in fields])
            words.append(word)
            array.append(embedding)
    np.save("/home/varun/treelstm-master/data/glove/words.npy",np.array(words))
    np.save("/home/varun/treelstm-master/data/glove/glove.npy",np.array(array))
    return np.array(words), np.array(array)
`

If you could point out any mistakes/errors and possible corrections.

Thanks

**Comment by ofirnachum** [comment]
Thanks for this info.  The issue with simple_demo is real and I've pushed a fix for it.

Your modified code for data_utils.read_embeddings_into_numpy looks strange - why do you set the word embedding like this? 

> embedding = np.array( [[float(ord(f)) for f in field] for field in fields])

First, the first element in fields (fields[0]) is a word, not a number.  Second, the embedding for a single word should be 1-dim, and here you make it 2-dim.  Third, by doing float(ord(f)) I don't think you are converting the numbers correctly - the numbers will end up being much bigger than they should be.

**Comment by varunsharma38** [comment]
While using the code snippet you suggested.

> embedding = np.array([float(f) for f in fields[1:]])

the  `array`  returned by the function is a 2D array of size 400000 but all the arrays are empty.

**Comment by varunsharma38** [comment]
Could you also explain the utility of the **tunable word embeddings** with respect to the overall code ?

**Comment by ofirnachum** [comment]
Tunable word embeddings improve accuracy.  Rather than initialize the embedding matrix randomly, initializing it to pre-trained word embeddings gives the training a "head-start".

I'm able to run the original code for read_embeddings_into_numpy without any problem:

`import data_utils`
`vocab = data_utils.Vocab()`
`vocab.load('../treelstm/data/sst/vocab-cased.txt')`
`words, embeddings = data_utils.read_embeddings_into_numpy('../treelstm/data/glove/glove.840B.300d.txt', vocab=vocab)
`

The shape of words should be (20693,) and the shape of embeddings should be (20693, 300). 

Perhaps you're using a different glove file?

**Comment by varunsharma38** [comment]
I am using glove.6B.300d.txt as my data file. However, I am getting both arrays as blank.

The shapes of word and embeddings (0,) and (0,) respectively.

**Comment by ofirnachum** [comment]
I see - there's likely something strange in the format of that file that is causing the issue then.  I'd recommend you either
1. use the 840B vectors,
2. train your model without the pre-trained word vectors,
3. or figure out what the formatting of the file is and write code to load it correctly.

**Comment by varunsharma38** [comment]
Thanks for your help

Can you just list the commands to be run after `fetch_and_preprocess` script along with the arguments to be used for 840B vectors.

What is the utility of this command:

> if [ ! -f $glove_dir/$glove_pre.$glove_dim.th ]; then
>     th scripts/convert-wordvecs.lua $glove_dir/$glove_pre.$glove_dim.txt \
>         $glove_dir/$glove_pre.vocab $glove_dir/$glove_pre.$glove_dim.th
> fi

 if we are not using the `.th` and `.vocab` files for `data_utils.read_embeddings_into_numpy` method.

Thanks in advance

**Comment by ofirnachum** [comment]
After running fetch_and_preprocess, you can run a python shell and enter the commands from my last comment to get the words and embeddings as numpy arrays.  Save them using

`import numpy as np`
`np.save('../treelstm/data/words.npy', words)`
`np.save('../treelstm/data/glove.npy', embeddings)`

Once this is done, you should be able to run python sentiment.py without issue.

The command you mention is from the treelstm git repo - it is probably not necessary for the sentiment code.

As you mentioned earlier, more detailed instructions in the readme would be helpful - I'll try to get around to this once I have more time.  For now, I'll close this issue.

**Comment by varunsharma38** [comment]
Hi

The code works fine now. Thanks a lot for your help.

**Comment by Jyoti1009** [comment]
Hi,

Kindly assist me on where do I get the vocab-cased.txt file.

Thanks in advance.

**Comment by ofirnachum** [comment]
You'll need to clone a separate repository and run the setup for it:

git clone https://github.com/stanfordnlp/treelstm.git
cd treelstm
./fetch_and_preprocess.sh
